### PR TITLE
Rate limited retry after

### PR DESCRIPTION
**Pull request by LorenzoCianciaruso**

### CODE REVIEW

File: pipe-http-server/src/integration/groovy/com/tesco/aqueduct/pipe/http/PipeReadControllerIntegrationSpec.groovy [hunk 1]
```diff
@@ -70,6 +74,30 @@ class PipeReadControllerIntegrationSpec extends Specification {
         "/pipe/1?location='someLocation'" | 200        | RETRY_AFTER_MS | "[]"
     }
 
+    //NOTE: the rate limiter when unused might give more permits than expected
+    //but the average on the long period should be respected
+    void "Rate limit retry after of 0ms when data is older than threshold"() {
```
Comment (ashish-sharma09) [inline pipe-http-server/src/integration/groovy/com/tesco/aqueduct/pipe/http/PipeReadControllerIntegrationSpec.groovy:79]: "retry-after" probably is more readable here

**Comment by LorenzoCianciaruso** [comment]
We are running a perf test to make sure that the rate limiter is working with proper load

This is what we are seeing in splunk:

https://splunk-operations.ocset.net/en-GB/app/SHC_tesco_app_price-service/search?earliest=1604595762&latest=1604596801&q=search%20index%3Dmagic_pipe_ppe%20sourcetype%3D%22kube%3Acontainer%3Aaqueduct-pipe%22%20level%3DDEBUG%20%22set%20retry%20time%20to%22%0A%7C%20rex%20field%3Dmessage%20%22set%20retry%20time%20to%20(%3F%3Cretry%3E%5Cd%7B1%2C5%7D)%22%0A%7C%20where%20retry%20%3D%200%0A%7C%20timechart%20span%3D1s%20count%20by%20retry&display.page.search.mode=fast&dispatch.sample_ratio=1&display.general.type=visualizations&display.page.search.tab=visualizations&sid=1604596287.66687_74A21D55-A4D8-4355-8B28-B10B72A77D9D

We expect a spike at the very beginning becaus the rate limiter will give out tokens when it's not used, and then the curve should smooth at 10rps.

We believe that logs are missing in the query above, but whenever we get logs, we can see that we have about 10rps issuing a 0 retry after.

Have a look at the query above and let us know your thoughts @ashish-sharma09  @adam-yorwerth

File: pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java [hunk 2]
```diff
@@ -54,6 +55,9 @@
     @Inject
     ContentEncoder contentEncoder;
 
+    @Inject
+    PipeRateLimiter rateLimiter;
```
Comment (ashish-sharma09) [inline pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java:59]: Can we have a constructor injection here?

File: pipe-http-server-cloud/src/main/resources/application.yml [hunk 3]
```diff
@@ -78,3 +78,5 @@ registry:
 compression:
   threshold-in-bytes: 1024
   logging: true
+
+rate-limiter.capacity: 10
```
Comment (ashish-sharma09) [inline pipe-http-server-cloud/src/main/resources/application.yml:82]: Is it worth configuring this in Gitops to be able to change and deploy quickly as needed?

File: pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java [hunk 4]
```diff
@@ -97,6 +102,12 @@
         return response;
     }
 
+    private boolean isBootstrapingAndCapacityAvailable(List<Message> messages) {
+        return messages.size() != 0
+            && messages.get(0).getCreated().isBefore(ZonedDateTime.now().minusHours(6))
+            && rateLimiter.tryAcquire(1);
```
Comment (ashish-sharma09) [inline pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java:108]: You can just use tryAcquire(), it is same as tryAcquire(1)

File: pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java [hunk 5]
```diff
@@ -97,6 +102,12 @@
         return response;
     }
 
+    private boolean isBootstrapingAndCapacityAvailable(List<Message> messages) {
+        return messages.size() != 0
```
Comment (ashish-sharma09) [inline pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java:106]: !messages.isEmpty()

File: pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java [hunk 6]
```diff
@@ -97,6 +102,12 @@
         return response;
     }
 
+    private boolean isBootstrapingAndCapacityAvailable(List<Message> messages) {
```
Comment (ashish-sharma09) [inline pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java:105]: Spelling typo - it's Bootstrapping

File: pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java (hunk 4 above)
Comment (adam-yorwerth) [reply pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java:108]: I think hardcoding with 1 is better, as it protects us from implementation changing in the future

File: pipe-http-server-cloud/src/main/resources/application.yml (hunk 3 above)
Comment (adam-yorwerth) [reply pipe-http-server-cloud/src/main/resources/application.yml:82]: Let's consider this as an enhancement to this feature, which can be made separately.

File: pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java (hunk 2 above)
Comment (adam-yorwerth) [reply pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java:59]: A good suggestion, but I think given the impact here and in dependencies we should do it as a separate change

File: pipe-http-server-cloud/build.gradle [hunk 7]
```diff
@@ -25,6 +25,7 @@ dependencies {
 
     implementation 'io.micronaut:micronaut-http-client'
     implementation 'com.sun.activation:javax.activation:1.2.0'
+    implementation 'com.google.guava:guava:29.0-jre'
```
Comment (adam-yorwerth) [inline pipe-http-server-cloud/build.gradle:28]: Guava 30 is released, so lets use that instead

File: pipe-http-server/build.gradle [hunk 8]
```diff
@@ -9,8 +9,9 @@ dependencies {
 
     testImplementation project(":pipe-storage-memory")
 
-    implementation('com.nixxcode.jvmbrotli:jvmbrotli:0.2.0')
-    implementation('com.nixxcode.jvmbrotli:jvmbrotli-linux-x86-amd64:0.2.0')
+    implementation 'com.nixxcode.jvmbrotli:jvmbrotli:0.2.0'
+    implementation 'com.nixxcode.jvmbrotli:jvmbrotli-linux-x86-amd64:0.2.0'
+    implementation 'com.google.guava:guava:29.0-jre'
```
Comment (adam-yorwerth) [inline pipe-http-server/build.gradle:14]: Guava 30

File: pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java [hunk 9]
```diff
@@ -97,6 +102,12 @@
         return response;
     }
 
+    private boolean isBootstrappingAndCapacityAvailable(List<Message> messages) {
+        return !messages.isEmpty()
+            && messages.get(0).getCreated().isBefore(ZonedDateTime.now().minusHours(6))
```
Comment (adam-yorwerth) [inline pipe-http-server/src/main/java/com/tesco/aqueduct/pipe/http/PipeReadController.java:107]: 6 hours should be in config, ideally as a duration `pipe.bootstrap.threshold: 6h` or something like that

**Review by adam-yorwerth — DISMISSED** [review]

**Review by adam-yorwerth — APPROVED** [review]

**Pull request merged**